llama.cpp

```diff
@@ -2504,7 +2504,11 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
                 }
             }
             if (tensor.name == "output.weight") {
-               new_type = GGML_TYPE_Q6_K;
+                int nx = tensor.ne.at(0);
+                int ny = tensor.ne.at(1);
+                if (nx % QK_K == 0 || ny % QK_K == 0) {
```

Comment:
Ahh, I don't think I'm really qualified to review your pulls!

The only thing I'd say is doing `if ((nx * ny) % QK_K == 0) {` in both places might be clearer for people with bad math knowledge (like me) that don't necessarily know if one of them is divisible then the number of elements in the tensor necessarily is.

There's currently no case for LLaMA models where the tensor sizes wouldn't be divisible. Correct?

Reply:
Thanks for spotting this! It must be `nx % QK_K == 0 && ny % QK_K == 0`

Reply:
Now I'm really confused. Was `&&` actually necessary here?

I assumed it was corrected initially, and did a little verification in the Python REPL:

```plaintext
>>> x = 256
>>> y = 256
>>> x * y % 256
0
>>> x = 13
>>> x * y % 256
0
>>> x = 256
>>> y = 13
>>> x * y % 256
0
```

Obviously that's not very scientific, but it seems like if one of the dimensions is divisible than the number of elements will be (at least for 2D tensors).

Reply:
Is is not about the total number of elements being a multiple of 256, but the number of columns being a multiple of 256. As a tensor can get used in left and right multiplications of a vector, it is best to check that rows and columns are a multiple of 256 (the current k-quants super=block size).

Reply:
This function call is always triggered with 1 row in a tensor multiplying a vector of embeddings, so `n` is the number of columns. Because the implementation does not handle the case where `n` (the number of columns in the quantized tensor) is not a multiple of `QK_K`, there is the assert.

Reply:
I guess the question I'm asking in a roundabout way is if some random code that uses GGML quantizes a tensor as Q6_K, Q4_K, whatever which _doesn't_ respect those invariants and then uses it in a matmul operation, what happens? Do we hit an assert and refuse to continue? It seems like the answer there would be "no" since Falcon 7B models appeared to work but didn't produce correct output.

Basically adding an assert to `llama.cpp` in that function does fix the problem for `llama.cpp` but it doesn't prevent other consumers of GGML + k-quants from shooting themselves in the foot unless there are asserts guarding the lower level functions.